### ISSUE TITLE
Sincere consult about the implementation of ADV

### ISSUE THREAD

**Issue by YiDongOuYang** [HaohanWang/HEX_experiments#2]
Thank you very much for your great work! 
However, I am a little bit confused about [this line](https://github.com/HaohanWang/HEX_experiments/blob/0012d59b27894adf98c61722f7c250942981310b/PACS/alex_cnn_top_adv.py#L192). In my point of view, this line only makes the classifier don't rely on spurious information to make prediction rather than "forcing the F_P to be invariant to F_G" mentioned in the paper.

(ADV aims to train F_P to be predictive for the semantic labels y while forcing the F_P to be invariant to F_G. I understand [this line](https://github.com/HaohanWang/HEX_experiments/blob/0012d59b27894adf98c61722f7c250942981310b/PACS/alex_cnn_top_adv.py#L186) achieves the goal of "training F_P to be predictive for the semantic labels y".)

I don't know whether I understand it correctly! Thank you very much for your reply!

**Comment by HaohanWang** [comment]
Hi, thanks for the nice words. Are you saying that "don't rely on" does not necessarily mean "be invariant to"?

**Comment by YiDongOuYang** [comment]
I really appreciate your swift reply!
To be honest, I don't understand why the F_P should "be equal to" (be invariant to) F_G since F_P is the logits obtained by using semantic representation while F_G is the logits obtained by using spurious information.

Could you please help me get the point that forcing the model not to rely on spurious information is equivalent to the invariant condition?
Thank you very much!

**Comment by YiDongOuYang** [comment]
It makes sense to me if we introduce a binary classifier telling us whether the representation is semantic or spurious, and then training the feature extractor by fooling this binary classifier(adopting the reverse gradient).

**Comment by HaohanWang** [comment]
Thanks for the suggestion. Let's assume we have feature matrices A and B, and functions f() and g(). 
I believe what you suggest is that we train an additional model h() for binary classification f(A) and g(B), and then train f() to fool h(). 
My code was to directly train f() so that g(B) cannot be predicted by f(A). I'm not sure if you notice the minus sign in the line you referred to.

**Comment by YiDongOuYang** [comment]
I get the point and thank you very much!

By the way, if I understand correctly, we actually want to maximize the discrepancy of g(B) and f(A), thus the model doesn't rely on spurious information. So, we force F_P to be as different as possible from F_G since F_P=f(A) and F_G=g(B). Why do you call it invariant?

**Comment by HaohanWang** [comment]
Thanks, good question. I don't think "invariance" has a rigorous definition even today. Back then, I just call "not be able to predict" invariance. It seems the implementation pushes further than "not be able to predict", then I guess there should be a better terminology to use.

**Comment by YiDongOuYang** [comment]
Thank you very much for your swift and clear illumination! I really appreciate it!
